### ISSUE TITLE
Product query extension is not executed for custom product model 

### ISSUE THREAD

**Issue by znikolovski** [adobe/aem-core-cif-components#504]
It looks like the first time the product query gets constructed is via the PageMetadata call in head.html on the CIF page component. The PageMetadata makes an adaptTo call to adapt the request to a product model which doesn't take in account the custom product model (which has the product query hook) resulting in AEM caching an incorrect product object. The subsequent call from the product component on the page doesn't result in a new product query call being made (which is correct from a performance perspective as we don't want to fetch the "same" product twice).

One way that I managed to get around this is to create a custom PageMetadata Sling model and overload the head.html HTL script to use it instead of the core one. 

### Expected Behaviour

New product attributes defined via a product query hook are available in the product component.

### Actual Behaviour

Calls to get methods for new attributes result in a NullPointerException as attributes are not available in the query response due to the query hook not being included.

#### Steps to Reproduce

1. Create a custom Product Model and add an extend* call in the @PostConstruct method
2. Add a get* method for the additional attributes
3. Extend product.html and include a call to any of the get methods

#### Platform and Version

CIF core components: 1.7.0
Magento Graphql: 6.0.0-magento235
AEM: Cloud Service SDK 2021.1.4830.20210128T075814Z-201217

#### Logs taken while reproducing problem

Caused by: java.lang.RuntimeException: java.lang.NullPointerException
        at com.shopify.graphql.support.AbstractResponse.lambda$converterWrapper$0(AbstractResponse.java:83) [com.adobe.commerce.cif.magento-graphql:6.0.0.magento235]
        at com.shopify.graphql.support.AbstractResponse.getAs(AbstractResponse.java:93) [com.adobe.commerce.cif.magento-graphql:6.0.0.magento235]
        at com.shopify.graphql.support.AbstractResponse.getAsString(AbstractResponse.java:106) [com.adobe.commerce.cif.magento-graphql:6.0.0.magento235]
        at com.CUSTOMER.aem.brands.core.models.commerce.CustomProductImpl.getMethod(CustomProductImpl.java:243) [breville-aem-brands.core:1.0.0.SNAPSHOT]
        ... 444 common frames omitted
Caused by: java.lang.NullPointerException: null
        at com.shopify.graphql.support.AbstractResponse.jsonAsString(AbstractResponse.java:197) [com.adobe.commerce.cif.magento-graphql:6.0.0.magento235]
        at com.shopify.graphql.support.AbstractResponse.lambda$converterWrapper$0(AbstractResponse.java:81) [com.adobe.commerce.cif.magento-graphql:6.0.0.magento235]
        ... 447 common frames omitted
25.02.2021 15:38:11.791 *INFO* [[0:0:0:

**Comment by cjelger** [comment]
@znikolovski Now that https://github.com/adobe/aem-core-cif-components/pull/507 has been merged, and based on https://github.com/adobe/aem-core-cif-components/pull/510 working fine with a product query extending the base product query, I think this issue is probably fixed. Can you please re-test with the current `-SNAPSHOT` version of the CIF components bundle?

**Comment by mhaack** [comment]
@znikolovski any updates on this?

**Comment by znikolovski** [comment]
Hey @mhaack sorry for the late reply - this is now resolved.

**Comment by Lawrence-M-Reynolds** [comment]
@znikolovski / @cjelger , are you sure this issue is fixed? I've been trying to extend the product functionality using the delegate pattern as suggested [here](https://experienceleague.adobe.com/docs/experience-manager-65/commerce/storefront/developing/customize-cif-components.html?lang=en). But I was still seeing the same issue while using core cif components version 2.3.1, which has the change for this ticket.

After going through the [pull request](https://github.com/adobe/aem-core-cif-components/pull/507/files) the only way I could see for the for it to load the custom query for the new productImpl class was to override the pagemetadata class and use my new product  interface to pass to the model factory on this line:

https://github.com/adobe/aem-core-cif-components/blob/1be13d82888ed09d399cd07c3f31f36cabf27470/bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/page/PageMetadataImpl.java#L71

Without this I was just seeing the same behavior as before.

**Comment by mhaack** [comment]
@Lawrence-M-Reynolds yes this should be fixed, custom models extending the core models should work.

What is the AEM & CIF version you are using?
Can you also share the custom model class? Especially the annotations of the class?

**Comment by Lawrence-M-Reynolds** [comment]
Thanks for responding @mhaack .

The versions are:
**AEM quickstart jar version:** 2021.8.5788.20210830T213613Z-210800
**cif:** Version 2021.10.5.00

I've attached the productimpl.java aswell as the PageMetadataImpl.java I had to create and reference in the page head template.
[CustomModels.zip](https://github.com/adobe/aem-core-cif-components/files/7469306/CustomModels.zip)

Maybe you can see something I've missed..

**Comment by mhaack** [comment]
Thanks for sharing the code. What are the extension you do in the `PageMetadataImpl.java` I could not spot any extra methods? If you don't add extra stuff here you don't need it. Also yours looks little older, which version of CIF Core Components do you have? 

For the product model you class annotations is not 100% and missing there "original" CIF core product interface. 

The annotation should look like

```
@Model(adaptables = SlingHttpServletRequest.class, adapters = { Product.class, CustomProductModel.class}, resourceType = CustomProductModelImpl.RESOURCE_TYPE)
```
Where `Product.class` is the CIF Core Component interface and `CustomProductModel.class` is your extended interface. With the class will be correctly picked up by our CIF core `PageMetadataImpl.java` and instantiated.

**Comment by Lawrence-M-Reynolds** [comment]
Hi @mhaack 
You pointed me in the right direction and I've managed to find out what the problem was. In [this tutorial](https://experienceleague.adobe.com/docs/experience-manager-65/commerce/storefront/developing/customize-cif-components.html?lang=en) they only set the custom model interface "MyProductTeaser" in the adaptable annotation. That was how I was doing it so it wasn't being picked up  in the PageMetadataImpl.java. But it is now that I've included the core interface as well:

`@Model(adaptables = SlingHttpServletRequest.class,
        adapters = { Product.class, com.adobe.cq.commerce.core.components.models.product.Product.class},
        resourceType = ProductImpl.RESOURCE_TYPE
)`

By the way, I named my interface and implementation class as "Product" and "ProductImpl" as well which is where the confusion was. And I didn't need to make any extensions to the PageMetadataImpl.java class, I was just using the new interface I'd created to pass to the model factory. It would probably have made sense to name it slightly different.

It would probably be a good idea to update the tutorial so that the core Product model interface is included as an adaptable for the custom model.

But thanks very much for your help @mhaack , it's made it much clearer to me now.

**Comment by mhaack** [comment]
Cool to hear you could fix the issue.

The original interface is only needed for components on the PDP or PLP, because here we have the page meta data which instantiates the Sling model already by using the original interface. That happens before your component gets called. You component then already gets the cache impl. and it is not instantiated a second time. Creating the model a second time cloud lead to the query a second time, which we want to avoid. We use [Sling model caching](https://sling.apache.org/documentation/bundles/models.html#caching-1) here.

For`MyProductTeaser` example this does not apply, hence here the tutorial is correct. But we should add this hint to the docs pages.

**Comment by mhaack** [comment]
Good catch @Lawrence-M-Reynolds you are right we are not following the recommendations here. The caching of the model was introduced later this is an oversight. We will fix it.

**Comment by cjelger** [comment]
Actually I was aware of the recommendation, but the main issue was fixed in sling models 1.4.10. Although the website says it may create issues, we've never had any in the past few years. Also note that this recommendation is pretty old and there have been quite a few changes in the sling caching since then, so I'm not 100% sure this still applies.
